src/dpr/utils/downloadUtils.ts

```diff
@@ -0,0 +1,64 @@
+import { Response, Request } from 'express'
+import { json2csv } from 'json-2-csv'
+import fs from 'fs-extra'
+import { Services } from '../types/Services'
+import Dict = NodeJS.Dict
+import logger from './logger'
+
+const convertToCsv = (reportData: Dict<string>[]) => {
+  const csvData = json2csv(reportData)
+  return csvData
+}
+
+const saveToFile = async (csvData: string, reportName: string, variantName: string) => {
+  const filepath = `./download/${reportName}-${variantName}-${new Date().toISOString()}.csv`
+  await fs.outputFile(filepath, csvData)
+  return {
+    filepath,
+  }
+}
+
+const applyColumnsAndSort = (data: Dict<string>[], columns: string[]) => {
+  return data.map((row) => {
+    return Object.keys(row)
+      .filter((key) => columns.includes(key))
+      .reduce((obj: Dict<string>, key) => {
+        // eslint-disable-next-line no-param-reassign
+        obj[key] = row[key]
+        return obj
+      }, {})
+  })
+}
+
+export default {
+  async downloadReport({ req, services, res }: { req: Request; services: Services; res: Response }) {
+    const userId = res.locals.user?.uuid ? res.locals.user.uuid : 'userId'
+    const token = res.locals.user?.token ? res.locals.user.token : 'token'
+
+    const { reportId, id, tableId, dataProductDefinitionsPath, reportName, variantName, cols: columns } = req.body
+
+    const canDownload = await services.downloadPermissionService.downloadEnabled(userId, reportId, id)
+    if (!canDownload) {
+      res.redirect(`/async/report/${reportId}/${id}/request/${tableId}/report/download-disabled`)
+    } else {
+      let reportData = await services.reportingService.getAsyncReport(token, reportId, id, tableId, {
```

Comment:
It also looks like summary and section reports will just download the raw source data. Is that an expected behaviour? Or should it be added to the assumptions?